plugins/inputs/intel_powerstat/intel_powerstat.go

```diff
@@ -0,0 +1,493 @@
+// +build linux
+
+package intel_powerstat
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/plugins/inputs"
+	"github.com/influxdata/telegraf/plugins/inputs/intel_powerstat/data"
+	"github.com/influxdata/telegraf/plugins/inputs/intel_powerstat/services"
+)
+
+const (
+	cpuC1                = "cpu_c1_state_residency"
+	cpuC6                = "cpu_c6_state_residency"
+	cpuBusyCycles        = "cpu_busy_cycles"
+	cpuBusyFrequency     = "cpu_busy_frequency"
+	cpuFrequency         = "cpu_frequency"
+	cpuTemperature       = "cpu_temperature"
+	percentageMultiplier = 100
+)
+
+var (
+	allowedProcessorModelsForC1C6 = []string{"0x37", "0x4D", "0x5C", "0x5F", "0x7A", "0x4C", "0x86", "0x96", "0x9C",
+		"0x1A", "0x1E", "0x1F", "0x2E", "0x25", "0x2C", "0x2F", "0x2A", "0x2D", "0x3A", "0x3E", "0x4E", "0x5E", "0x55", "0x8E",
+		"0x9E", "0x6A", "0x6C", "0x7D", "0x7E", "0x9D", "0x3C", "0x3F", "0x45", "0x46", "0x3D", "0x47", "0x4F", "0x56",
+		"0x66", "0x57", "0x85", "0xA5", "0xA6", "0x8F", "0x8C", "0x8D"}
+)
+
+// PowerStat plugin enables monitoring CPU utilization and power consumption on Intel platform.
+type PowerStat struct {
+	CPUMetrics []string        `toml:"cpu_metrics"`
+	Log        telegraf.Logger `toml:"-"`
+
+	fs   services.FileService
+	rapl services.RaplService
+	msr  services.MsrService
+
+	perCoreMetrics         bool
+	coreCPUFrequency       bool
+	c1CoreStateResidency   bool
+	c6CoreStateResidency   bool
+	processorBusyCycles    bool
+	coreCPUTemperature     bool
+	processorBusyFrequency bool
+	cpuInfo                map[string]*data.CPUInfo
+	skipFirstIteration     bool
+}
+
+// Init performs one time setup of the plugin.
+func (p *PowerStat) Init() error {
+	p.ParseCPUMetricsConfig()
+	err := p.verifyProcessor()
+	if err != nil {
+		return err
+	}
+	// Set perCoreMetrics flag only when there is at least one core metric enabled.
+	if p.c1CoreStateResidency || p.c6CoreStateResidency || p.processorBusyCycles || p.coreCPUTemperature ||
+		p.processorBusyFrequency || p.coreCPUFrequency {
+		p.perCoreMetrics = true
```

Comment:
This is in the later code just a guard against `p.msr` being `nil`. How about directly checking for `p.msr` being nil and getting rid of `p.perCoreMetrics` altogether. It was hard to me to read the later checks because I had to remember that `p.msr` cannot be nil if  `p.perCoreMetrics` is true. Checking directly for nil would have been easier to read.

Reply:
You're right. Corrected.